src/loaders/compute_tools/tool_common.py

```diff
@@ -298,6 +298,7 @@ def parallel_single_execution(self, tool_callable: Callable[[str, str, Path, Pat
                  meta.get(loader_common_names.META_UNCOMPRESSED_FILE,
                           meta[loader_common_names.META_SOURCE_FILE]),
                  output_dir,
+                 self._program_threads,
```

Comment:
Doesn't this make the thread count different for single and batch run programs? Batch programs will never use more than `_threads` threads since they divide that by `_program_threads` to get the instance count IIUC. Single program threads will now use `_threads` * `_program_threads` threads

Reply:
I will need to unify the logic for single and batch programs. 

For batch program, I designed to compute the threads number by _threads/_program_threads. However, later we decided to have Taskfarmer handle parallelization. 
https://github.com/kbase/collections/blob/develop/src/loaders/jobs/taskfarmer/task_generator.py#L208-L209

For single program, total threads are _threads * _program_threads * NODE_THREADS. When I schedule the job, I manually set NODE_THREADS to 1. 

Does that make sense?

Reply:
> threads number by _threads/_program_threads.

you mean instances number?

> For single program, total threads are _threads * _program_threads * NODE_THREADS. When I schedule the job, I manually set NODE_THREADS to 1.

I'm kind of lost here - is NODE_THREADS a slurm variable?

Reply:
yea. `NODE_THREADS` is a slurm variable that controls how many threads each node can run. We set it to 4 now. Because it gives us the best performance for gtdb_tk. 

And that's why we set _program_threads == _threads
https://github.com/kbase/collections/blob/develop/src/loaders/jobs/taskfarmer/task_generator.py#L208-L209

Reply:
Are those actual OS threads or does it really mean jobs / node?

Reply:
OS threads. For example, if we set `NODE_THREADS` to 2, and there are 6 tasks, it will run 2 tasks at a time.

Reply:
when I run eggnog, I manually change NODE_THREADS to 1.

Reply:
And therefore those tasks will only get one os thread each? In that case what's the point of specifying threads per program at all?

Reply:
> when I run eggnog, I manually change NODE_THREADS to 1.

If NODE_THREADS == os threads then all the tests you did trying different thread counts per program are invalid since the 1 instance only got one os thread

Reply:
This can't be right I think we must be talking past each other

Reply:
oh no. that's not what I meant. `The THREADS variable controls the number of tasks per node`

https://docs.nersc.gov/jobs/workflow/taskfarmer/#step-3-create-a-batch-script-submit_taskfarmersl

Reply:
Ok, good. So for a batch job _theads is divided by _program_threads, which are equal, and therefore one instance is run per NODE_THREAD, with the instance using _threads os threads via some CLI argument.

For a single job, _threads instances are run per NODE_THREAD and each instance uses _program_threads os threads via a CLI argument, which is why you set NODE_THREADS to 1.

Right?

Reply:
yea. That's correct. I haven't fully thought through how to streamline the logic yet, but I'll make some improvements in task_generator.py to enhance clarity.

Reply:
yeah, you and I should talk through a refactor before we start anything , it's a pretty odd setup at this point.

I wouldn't bother with it for now unless you have cycles other than maybe adding some docs in an appropriate place explaining how it works now

Reply:
cool. Sounds good. I will add some notes in task_generator in my next PR. I need to do some modification there as well.

Reply:
:+1: